File: peterson.c
/* Deterministic contract lowered from the reviewed V2 domain 'peterson'.
 * Human review of the reviewed artifact is required before trust. */

typedef struct {
    int flag0;
    int flag1;
    int turn;
    int pc0;
    int pc1;
} peterson;

/*@
  requires \valid(counter);
  assigns counter->flag0, counter->flag1, counter->turn, counter->pc0, counter->pc1;
  ensures counter->flag0 == 0 && counter->flag1 == 0 && counter->turn == 0 && counter->pc0 == 0 && counter->pc1 == 0;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_init(peterson *counter) {
    counter->flag0 = 0;
    counter->flag1 = 0;
    counter->turn = 0;
    counter->pc0 = 0;
    counter->pc1 = 0;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->flag0;
*/
int peterson_get_flag0(const peterson *counter) {
    return counter->flag0;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->flag1;
*/
int peterson_get_flag1(const peterson *counter) {
    return counter->flag1;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->turn;
*/
int peterson_get_turn(const peterson *counter) {
    return counter->turn;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->pc0;
*/
int peterson_get_pc0(const peterson *counter) {
    return counter->pc0;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->pc1;
*/
int peterson_get_pc1(const peterson *counter) {
    return counter->pc1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
  requires counter->pc0 == 0;
  assigns counter->flag0, counter->turn, counter->pc0;
  ensures counter->flag0 == 1 && counter->turn == 1 && counter->pc0 == 1;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_request0(peterson *counter) {
    counter->flag0 = 1;
    counter->turn = 1;
    counter->pc0 = 1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
  requires counter->pc0 == 1;
  requires (counter->flag1 == 0) || (counter->turn == 0);
  assigns counter->pc0;
  ensures counter->pc0 == 2;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_enter0(peterson *counter) {
    counter->pc0 = 2;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
  requires counter->pc0 == 2;
  assigns counter->flag0, counter->pc0;
  ensures counter->flag0 == 0 && counter->pc0 == 0;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_exit0(peterson *counter) {
    counter->flag0 = 0;
    counter->pc0 = 0;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
  requires counter->pc1 == 0;
  assigns counter->flag1, counter->turn, counter->pc1;
  ensures counter->flag1 == 1 && counter->turn == 0 && counter->pc1 == 1;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_request1(peterson *counter) {
    counter->flag1 = 1;
    counter->turn = 0;
    counter->pc1 = 1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
  requires counter->pc1 == 1;
  requires (counter->flag0 == 0) || (counter->turn == 1);
  assigns counter->pc1;
  ensures counter->pc1 == 2;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_enter1(peterson *counter) {
    counter->pc1 = 2;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
  requires counter->pc1 == 2;
  assigns counter->flag1, counter->pc1;
  ensures counter->flag1 == 0 && counter->pc1 == 0;
  ensures (0 <= counter->flag0) && (counter->flag0 <= 1) && (0 <= counter->flag1) && (counter->flag1 <= 1) && (0 <= counter->turn) && (counter->turn <= 1) && (0 <= counter->pc0) && (counter->pc0 <= 2) && (0 <= counter->pc1) && (counter->pc1 <= 2) && (!((counter->pc0 == 2) && (counter->pc1 == 2))) && ((counter->pc0 == 2) ==> (counter->flag0 == 1)) && ((counter->pc1 == 2) ==> (counter->flag1 == 1)) && ((counter->pc0 == 2) ==> ((counter->flag1 == 0) || (counter->turn == 0))) && ((counter->pc1 == 2) ==> ((counter->flag0 == 0) || (counter->turn == 1))) && ((counter->pc0 == 1) ==> (counter->flag0 == 1)) && ((counter->pc1 == 1) ==> (counter->flag1 == 1));
*/
void peterson_exit1(peterson *counter) {
    counter->flag1 = 0;
    counter->pc1 = 0;
}

File: Peterson.rs
use prusti_contracts::*;

pub struct Peterson {
    pub flag0: i32,
    pub flag1: i32,
    pub turn: i32,
    pub pc0: i32,
    pub pc1: i32,
}

impl Peterson {
    #[ensures(result.flag0 == 0 && result.flag1 == 0 && result.turn == 0 && result.pc0 == 0 && result.pc1 == 0 && (0 <= result.flag0) && (result.flag0 <= 1) && (0 <= result.flag1) && (result.flag1 <= 1) && (0 <= result.turn) && (result.turn <= 1) && (0 <= result.pc0) && (result.pc0 <= 2) && (0 <= result.pc1) && (result.pc1 <= 2) && (!((result.pc0 == 2) && (result.pc1 == 2))) && ((result.pc0 == 2) ==> (result.flag0 == 1)) && ((result.pc1 == 2) ==> (result.flag1 == 1)) && ((result.pc0 == 2) ==> ((result.flag1 == 0) || (result.turn == 0))) && ((result.pc1 == 2) ==> ((result.flag0 == 0) || (result.turn == 1))) && ((result.pc0 == 1) ==> (result.flag0 == 1)) && ((result.pc1 == 1) ==> (result.flag1 == 1)))]
    pub fn new() -> Self {
        Self { flag0: 0, flag1: 0, turn: 0, pc0: 0, pc1: 0 }
    }

    #[pure]
    #[ensures(result == self.flag0)]
    pub fn get_flag0(&self) -> i32 {
        self.flag0
    }

    #[pure]
    #[ensures(result == self.flag1)]
    pub fn get_flag1(&self) -> i32 {
        self.flag1
    }

    #[pure]
    #[ensures(result == self.turn)]
    pub fn get_turn(&self) -> i32 {
        self.turn
    }

    #[pure]
    #[ensures(result == self.pc0)]
    pub fn get_pc0(&self) -> i32 {
        self.pc0
    }

    #[pure]
    #[ensures(result == self.pc1)]
    pub fn get_pc1(&self) -> i32 {
        self.pc1
    }

    #[requires((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    #[requires(self.pc0 == 0)]
    #[ensures(self.flag0 == 1 && self.turn == 1 && self.pc0 == 1)]
    #[ensures((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    pub fn request0(&mut self) {
        self.flag0 = 1;
        self.turn = 1;
        self.pc0 = 1;
    }

    #[requires((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    #[requires(self.pc0 == 1)]
    #[requires((self.flag1 == 0) || (self.turn == 0))]
    #[ensures(self.pc0 == 2)]
    #[ensures((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    pub fn enter0(&mut self) {
        self.pc0 = 2;
    }

    #[requires((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    #[requires(self.pc0 == 2)]
    #[ensures(self.flag0 == 0 && self.pc0 == 0)]
    #[ensures((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    pub fn exit0(&mut self) {
        self.flag0 = 0;
        self.pc0 = 0;
    }

    #[requires((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    #[requires(self.pc1 == 0)]
    #[ensures(self.flag1 == 1 && self.turn == 0 && self.pc1 == 1)]
    #[ensures((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    pub fn request1(&mut self) {
        self.flag1 = 1;
        self.turn = 0;
        self.pc1 = 1;
    }

    #[requires((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    #[requires(self.pc1 == 1)]
    #[requires((self.flag0 == 0) || (self.turn == 1))]
    #[ensures(self.pc1 == 2)]
    #[ensures((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    pub fn enter1(&mut self) {
        self.pc1 = 2;
    }

    #[requires((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    #[requires(self.pc1 == 2)]
    #[ensures(self.flag1 == 0 && self.pc1 == 0)]
    #[ensures((0 <= self.flag0) && (self.flag0 <= 1) && (0 <= self.flag1) && (self.flag1 <= 1) && (0 <= self.turn) && (self.turn <= 1) && (0 <= self.pc0) && (self.pc0 <= 2) && (0 <= self.pc1) && (self.pc1 <= 2) && (!((self.pc0 == 2) && (self.pc1 == 2))) && ((self.pc0 == 2) ==> (self.flag0 == 1)) && ((self.pc1 == 2) ==> (self.flag1 == 1)) && ((self.pc0 == 2) ==> ((self.flag1 == 0) || (self.turn == 0))) && ((self.pc1 == 2) ==> ((self.flag0 == 0) || (self.turn == 1))) && ((self.pc0 == 1) ==> (self.flag0 == 1)) && ((self.pc1 == 1) ==> (self.flag1 == 1)))]
    pub fn exit1(&mut self) {
        self.flag1 = 0;
        self.pc1 = 0;
    }
}
